Remove recommendation for WebRequest (#30305)
dotnet/docs: docs/framework/network-programming/best-practices-for-system-net-classes.md

---
title: "Best Practices for System.Net Classes"
description: Follow these recommendations to use the classes contained in System.Net to their best advantage in .NET Framework programming.
ms.date: 07/21/2022
helpviewer_keywords: 
  - "sending data, best practices"
  - "requesting data from internet, best practices"
  - "WebRequest class, best practices"
  - "data requests, best practices"
  - "WebResponse class, best practices"
  - "best practices, data requests"
  - "receiving data, best practices"
ms.assetid: 716decc6-5952-47b7-9c5a-ba6fc5698684
---
# Best practices for System.Net classes

The following recommendations will help you use the classes contained in <xref:System.Net> to their best advantage:  
  
- For Transport Layer Security (TLS) best practices, see [Transport Layer Security (TLS) best practices with .NET Framework](tls.md).

> - Use <xref:System.Net.Http.HttpClient> to send HTTP requests instead of <xref:System.Net.WebRequest>, which was obsoleted in .NET 6. In .NET Framework, you can create a new `HttpClient` instance each time you need to send a request.

Reviewer: ```suggestion
- Use <xref:System.Net.Http.HttpClient> to send HTTP requests instead of <xref:System.Net.WebRequest>, which was obsoleted in .NET 6. In .NET Framework, you should try to create and reuse a single `HttpClient` instance each time you need to send a request.
```

We might want to update this one too:

https://docs.microsoft.com/dotnet/fundamentals/networking/httpclient-guidelines

We could probably link to this doc too:

- https://docs.microsoft.com/azure/architecture/antipatterns/improper-instantiation/

Author: I'm not sure - [this comment](https://github.com/dotnet/runtime/issues/68866#issuecomment-1119916486) suggests the .NET Framework implementation of HttpClient is different and disposing it doesn't impact connection management. @wfurt  Can you take a look at this PR perhaps?

Reviewer: I'm somewhat familiar with that, but fundamentally; we're introducing two different patterns for essentially the same type. The key difference between them is their target framework, either .NET Framework or .NET in this case. My concern is that it's risky to try to teach learners of .NET two ways to use a single type depending on the TFM:

1. What happens when a developer from .NET Framework updates to .NET 6, ideally, they wouldn't have to change the pattern of their consumption of the `HttpClient`.

2. My understanding is that there isn't any harm in reusing an `HttpClient` instance in .NET Framework, so while it's probably cleaner to wrap it in a `using` is it really necessary here — I'm not sure, but would love to know?

Does that make sense?

Author: @IEvangelist The [PooledConnectionLifetime property](https://docs.microsoft.com/en-us/dotnet/api/system.net.http.socketshttphandler.pooledconnectionlifetime?view=net-6.0#system-net-http-socketshttphandler-pooledconnectionlifetime) isn't available in .NET Framework. So wouldn't users then run into the DNS changes issue?

> <xref:System.Net.Http.HttpClient> only resolves DNS entries when a connection is created. It does not track any time to live (TTL) durations specified by the DNS server. If DNS entries change regularly, which can happen in some container scenarios, the client won't respect those updates.
